core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/progress/listener/ProgressUpdater.java

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.progress.listener;
+
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.reactivestreams.Subscriber;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.SdkRequest;
+import software.amazon.awssdk.core.SdkResponse;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.core.async.listener.AsyncRequestBodyListener;
+import software.amazon.awssdk.core.internal.progress.ProgressListenerContext;
+import software.amazon.awssdk.core.internal.progress.ProgressListenerFailedContext;
+import software.amazon.awssdk.core.internal.progress.snapshot.DefaultProgressSnapshot;
+import software.amazon.awssdk.core.progress.listener.SdkRequestProgress;
+import software.amazon.awssdk.core.progress.snapshot.ProgressSnapshot;
+
+/**
+ * ProgressUpdater exposes methods that invokes listener methods to update and store request progress state
+ */
+@SdkInternalApi
+public class ProgressUpdater {
+    private final DefaultSdkRequestProgress uploadProgress;
+    private final DefaultSdkRequestProgress downloadProgress;
+    private final ProgressListenerContext context;
+    private final ProgressListenerInvoker listenerInvoker;
+    private final CompletableFuture<Void> endOfStreamFuture;
+
+    public ProgressUpdater(SdkRequest sdkRequest,
+                           Long contentLength) {
+        DefaultProgressSnapshot.Builder uploadProgressSnapshotBuilder = DefaultProgressSnapshot.builder();
+        uploadProgressSnapshotBuilder.transferredBytes(0L);
+        Optional.ofNullable(contentLength).ifPresent(uploadProgressSnapshotBuilder::totalBytes);
+
+        ProgressSnapshot uploadProgressSnapshot = uploadProgressSnapshotBuilder.build();
+        uploadProgress = new DefaultSdkRequestProgress(uploadProgressSnapshot);
+
+        DefaultProgressSnapshot.Builder downloadProgressSnapshotBuilder = DefaultProgressSnapshot.builder();
+        downloadProgressSnapshotBuilder.transferredBytes(0L);
+        Optional.ofNullable(contentLength).ifPresent(downloadProgressSnapshotBuilder::totalBytes);
```

Comment:
When will we update totalbytes?

Reply:
When the ProgressUpdater is initialized, as in line 177 in progressUpdaterTest, 

```
ProgressUpdater progressUpdater = new ProgressUpdater(sdkRequest, contentLength);
```